### DISCUSSION TITLE
FreeU

### DISCUSSION THREAD

**Discussion by 3Diva** [lllyasviel/Fooocus#618] [Show and tell]
I don't know EXACTLY what it is (I read a little bit about it here: https://github.com/ChenyangSi/FreeU/tree/main ), but I thought I'd go ahead and test it out and document want it looks like when applied. For this test I used the same seed and prompt and disabled the the styles options to assure that nothing else is effecting the generated image. I'm using the checkpoint model CrystalClear XL with no refiner and no lora.

I'm making BIG number value changes - using large number increments so that I can better see how each setting effects the image.

Base image with FreeU Not enabled:
![2023-10-09_22-07-32_4537](https://github.com/lllyasviel/Fooocus/assets/51475772/3c964311-11f0-45ae-9166-346bb4aee29e)

FreeU enabled, with default settings of B1 = 1.02, B2 = 1.02, S1 = 0.99, S2 = 0.95:
![2023-10-09_22-09-57_1652](https://github.com/lllyasviel/Fooocus/assets/51475772/97174769-d71c-47b7-a1e8-b1a08301c8f0)

FreeU with B1 = 2 (other settings default):
![2023-10-09_22-13-13_3577](https://github.com/lllyasviel/Fooocus/assets/51475772/3c4247ba-2336-40dc-a653-62058e897315)

FreeU with B2 = 2 (other settings default):
![2023-10-09_22-14-56_8336](https://github.com/lllyasviel/Fooocus/assets/51475772/a0d9eba7-db83-4f6f-b04b-9f753974cee4)

FreeU with S1 = 1.5 (other settings default):
![2023-10-09_22-45-46_8203](https://github.com/lllyasviel/Fooocus/assets/51475772/15aeb177-7309-4348-906d-144ff3b6f181)

FreeU with S1 = 2 (other settings default):
![2023-10-09_22-47-56_9853](https://github.com/lllyasviel/Fooocus/assets/51475772/6bd730e3-a562-4ff5-b3b4-f5034c1ba943)

FreeU with S1 = 2.5 (other settings default):
![2023-10-09_22-49-39_7880](https://github.com/lllyasviel/Fooocus/assets/51475772/c35d828b-900c-4696-aa77-f35998d8dd21)

FreeU with S1 = 3 (other settings default):
![2023-10-09_22-51-31_3378](https://github.com/lllyasviel/Fooocus/assets/51475772/b6fab7c9-14b7-4a7b-9a20-d648899851bf)

FreeU with S1 = 3.5 (other setting default):
![2023-10-09_22-52-56_2234](https://github.com/lllyasviel/Fooocus/assets/51475772/68e5b3c2-b054-4393-8a5c-371472ab92dc)

FreeU with S1 = 4 (other settings default):
![2023-10-09_22-55-07_8680](https://github.com/lllyasviel/Fooocus/assets/51475772/f6df5e8b-c5c0-4b5f-9666-fad913576185)
 
FreeU with S2 = 1.5 (other settings default):
![2023-10-09_22-56-59_8460](https://github.com/lllyasviel/Fooocus/assets/51475772/a16d618a-0e93-41bd-9900-af604b1fa4a2)

FreeU with S2 = 2 (other settings default):
![2023-10-09_22-58-22_7373](https://github.com/lllyasviel/Fooocus/assets/51475772/131baa80-eafa-4409-a1fd-c2990e5834a4)

FreeU with S2 = 2.5 (other settings default):
![2023-10-09_23-00-01_5958](https://github.com/lllyasviel/Fooocus/assets/51475772/8e6ad053-43f8-4e78-8fe4-cc97cdbb62d4)

FreeU with S2 = 3 (other settings default):
![2023-10-09_23-01-30_4276](https://github.com/lllyasviel/Fooocus/assets/51475772/028a2e1b-18a0-4085-9e90-65ea9ae1d635)

FreeU with S2 = 3.5 (other settings default):
![2023-10-09_23-02-57_9949](https://github.com/lllyasviel/Fooocus/assets/51475772/b741c66a-04ef-419f-a63c-01d146f83e3c)

FreeU with S2 = 4 (other settings default):
![2023-10-09_23-04-34_8146](https://github.com/lllyasviel/Fooocus/assets/51475772/1cbeb5ef-f339-471b-b10c-4c85a4817e61)

I REALLY like the results with the default FreeU settings. I think that FreeU improves the images quite nicely!

**Comment by 3Diva** [comment]
Second Round of Testing. This time we're going BELOW the base settings. As above I'll be only changing one setting at a time and leaving the rest on their default settings.

B1 = 0.5:
![2023-10-09_23-33-04_5405](https://github.com/lllyasviel/Fooocus/assets/51475772/59174a80-cd0e-492b-8c13-b92d94adde74)

B1 = 0.25:
![2023-10-09_23-36-41_1991](https://github.com/lllyasviel/Fooocus/assets/51475772/6b3df35f-779c-4ff3-b9cd-75cae23c6dbc)

B2 = 0.5:
![2023-10-09_23-38-12_2645](https://github.com/lllyasviel/Fooocus/assets/51475772/d87a53c0-c719-481c-896f-f48fad6296ca)

B2 = 0.25:
![2023-10-09_23-59-37_8180](https://github.com/lllyasviel/Fooocus/assets/51475772/da04f05b-0bf6-4460-8f57-82553e41084a)

S1 = 0.5:
![2023-10-09_23-58-07_3356](https://github.com/lllyasviel/Fooocus/assets/51475772/b3fd2795-3e47-46ee-8eeb-8e4eee2f32aa)

S1 = 0.25:
![2023-10-10_00-01-08_4683](https://github.com/lllyasviel/Fooocus/assets/51475772/ebd4ed89-2b10-42ed-b123-3767640e496b)

S2 = 0.5:
![2023-10-10_00-02-50_1898](https://github.com/lllyasviel/Fooocus/assets/51475772/7bdd37a8-2b97-421e-a3b6-db5fbc1b19f5)

S2 = 0.25:
![2023-10-10_00-04-19_6085](https://github.com/lllyasviel/Fooocus/assets/51475772/30aa0ba9-f7bc-4016-ace2-a4837f675a40)

S2 = 0:
![2023-10-10_00-05-53_5725](https://github.com/lllyasviel/Fooocus/assets/51475772/1f509b19-82b8-4c53-8885-21fd4736e6dd)

**Comment by 3Diva** [comment]
Here are some settings combos I ended up really liking for this image.

![Screenshot 2023-10-10 003749](https://github.com/lllyasviel/Fooocus/assets/51475772/f9cece2b-d3b8-4815-b235-e9e4758a6a94)
![2023-10-10_00-35-54_9114](https://github.com/lllyasviel/Fooocus/assets/51475772/74a55f3e-cfbb-4a4f-8b8d-c1193e563329)

![Screenshot 2023-10-10 004046](https://github.com/lllyasviel/Fooocus/assets/51475772/603128fe-102b-4055-ab95-545971bda99f)
![2023-10-10_00-40-22_6405](https://github.com/lllyasviel/Fooocus/assets/51475772/bb31f780-d998-46fe-8271-0b367f2e8ad4)

![Screenshot 2023-10-10 005002](https://github.com/lllyasviel/Fooocus/assets/51475772/e194fa80-ac7e-46d3-a653-17684b059ac2)
![2023-10-10_00-49-07_4673](https://github.com/lllyasviel/Fooocus/assets/51475772/443ab50a-0d24-4269-9f12-cfc1e1aee2e6)

![Screenshot 2023-10-10 011206](https://github.com/lllyasviel/Fooocus/assets/51475772/c8d47b08-6eed-44df-b4e6-7bb1044d9081)
![2023-10-10_01-11-33_4197](https://github.com/lllyasviel/Fooocus/assets/51475772/7e74ba4a-6645-4a15-a7c8-1b3acaebde98)


I'll have to do further testing to see how the settings look with different images.

**Reply by AkoZ** [reply]
About FreeU : https://github.com/ChenyangSi/FreeU which is already (yesterday ?) included in FoOocus ! ;)
Copying here some [youtube FreeU](https://www.youtube.com/watch?v=8XJH6uZjNzA&t=297s) comments : 
* the larger blobs are B values control the style and image quality and values around 1 for both give good results.
* tiny details and textures are the S-values control the quality/number (e.g. number of legs of an animal) of the objects on the image
** best results with values around and higher for S1 and with low values between 0-0.3 for S2.
** When the distance between S1 and S2 becomes smaller (S1~= S2), more and more objects are generated.
*  Normally the S and B interplay in a way that tends to find balance the high (S) and low (B) frequencies of the image.

**Reply by 3Diva** [reply]
Thank you for the great info! Could you clarify what this means: "best results with values around and higher for S1"?

**Reply by AkoZ** [reply]
* just that S1 is > S2 but very near ;) as 0.5 if S2=0.3 but that stays as theory as not so viewable.
* S textured lines are mitigating B (smoother colored)
* * seen in article, well the B=S=1 seems really significant in total balanced scene - then affect more about (B) flat mass and (S) textures, then after those factors are applied to amplify the details S or the semantic whole B.
 
* here is the paper: https://arxiv.org/pdf/2309.11497.pdf _and many here are citations from it !_ better understanding about low and high frequencies is described as some "High" seems like (controlnet canny ones)/ (** S = sketches** as lines (but named **skip feature scaling factors**) and then the flat or well to say the fullness is said about **B= Blobs ( named backbone feature factors** inside the lines is the colors and bold...)
* this is essential: "the diffusion models are comprised of the diffusion process and **the denoising process**. During the diffusion process, Gaussian noise is gradually added to the input data and eventually corrupts it into approximately pure Gaussian
noise. During the denoising process, the original input data is recovered from its noise state through a learned sequence
of inverse diffusion operations. Usually, a U-Net is trained to iteratively predict the noise to be removed at each denoising step."
![image](https://github.com/lllyasviel/Fooocus/assets/449844/79cae5c3-b2fe-41b9-b094-5d3a531fdb6f)
-- a " progressive denoising process, showcasing the generated images across successive iteration"
__ B: a " gradual modulation of **low frequency components, exhibiting a subdued rate of change**,
__ B: "global layouts and smooth color. These components encapsulate the foundational global elements that constitute the image’s essence and representation"
__ B: " the **backbone feature factors**, aims to amplify the feature maps of the main backbone, thereby bolstering the
denoising process. However, we find that while the inclusion of backbone feature scaling factors yields significant
improvements, it can occasionally lead to an undesirable oversmoothing of textures. 
* To mitigate this issue is introduced the second factor, **skip feature scaling factors**, aiming
to alleviate the problem of texture oversmoothing
__ S: "..while their high-frequency components display more pronounced dynamics throughout the denoising process"
__ S: ".. rapid changes in the images, such as edges and textures. These finer details are markedly sensitive to noise, often
manifesting as random high-frequency information whennoise is introduced to an image. Consequently, denoising processes need to expunge noise while upholding indispensable intricate details
* " **FreeU**, which substantially enhances the sample quality of diffusion models without incurring any additional computational costs. Motivated by the fundamental **role played by both skip connections and backbone features in U-Net architectures**, we conduct
an in-depth analysis of their **effects in diffusion U-Net**.
       Our investigation reveals that the primary backbone (B-factor) primarily contributes to denoising, while the skip connections (S-factor) predominantly introduce high-frequency features into the decoder, potentially leading to a neglect of essential backbone semantics. 
       To address this, we strategically re-weight the contributions originating from the U-Net’s skip connections and backbone feature maps. This re-weighting process capitalizes on the unique strengths of both U-Net components,
resulting in a substantial improvement in sample quality across a wide range of text prompts and random seeds.
* well it seems that when they pass other "filters" as the controlnet ones they will find other "moderating" factors then.. as here is the only semantic (B flat bold Blobs) and the S which is not precise as a controlnet HED or MLSD .. precisions will be found early now.

**Comment by lllyasviel** [comment]
freeu has some unknow numerial precision problems with AMD or MAC devices so it cannot be set as default.
also it has same problem as negative adm that it sometimes mess small objects like human eyes ( especially when eyes are small, eg full body images
this can be solved by early stopping of FreeU (and, negative adm already has early stopping now so it will not mess small objects and eyes, in theory)
but FreeU does not have early stopping and I am going to write it if the numerial precision problems with AMD or MAC can be solved and  when it is safe to use it

**Reply by AkoZ** [reply]
@lllyasviel thank YOU so much for your great participâtion to the stable diffusion tools ! Great one you are !

**Comment by Abracadd** [comment]
And where to upload an image to use the freeU   mode ?
